### PR TITLE
Support loading prop on header cells

### PR DESCRIPTION
**Pull request by michael-yx-wu**
#### Checklist
- [x] Works with EditableName
- [x] Works with children content (spacing only)
  - Users will be responsible with using the `pt-skeleton` class on their header cell children
- [x] Update documentation
- [x] Includes tests

### CODE REVIEW

File: packages/table/src/common/loadableContent.tsx [hunk 1]
```diff
@@ -29,7 +29,7 @@ export class LoadableContent extends React.Component<ILoadableContentProps, {}>
 
     public constructor(props: ILoadableContentProps) {
         super(props);
-        this.calculateStyle(props.loading);
+        this.calculateStyle(props.variableLength);
```
Comment (giladgray) [inline packages/table/src/common/loadableContent.tsx]: preferable to have this function return the styles instead of setting them internally (#sideeffect).

`this.style = this.calculateStyle()`

File: packages/table/src/headers/columnHeaderCell.tsx [hunk 2]
```diff
@@ -59,6 +62,15 @@ export interface IColumnHeaderCellProps extends IColumnNameProps, IProps {
     isColumnSelected?: boolean;
 
     /**
+     * If true, the column name (string or `ReactElement`) will be replaced with a fixed-height
```
Comment (giladgray) [inline packages/table/src/headers/columnHeaderCell.tsx]: paren clause is not helpful, please remove. say "column `name`" to reference the prop, from which the user can infer the typings.

File: packages/table/src/headers/columnHeaderCell.tsx [hunk 3]
```diff
@@ -125,17 +137,19 @@ export class ColumnHeaderCell extends React.Component<IColumnHeaderCellProps, IC
...
             "bp-table-header-selected": isColumnSelected,
+            [Classes.LOADING]: loading,
         }, className);
+
         return (
             <div className={classes} style={style}>
                 {this.renderName()}
                 {this.maybeRenderContent()}
-                {resizeHandle}
+                {loading ? null : resizeHandle}
```
Comment (giladgray) [inline packages/table/src/headers/columnHeaderCell.tsx]: use `undefined` instead of `null` in cases like this please

Comment (giladgray) [inline packages/table/src/headers/columnHeaderCell.tsx]: comments in this file also apply to file below

**Comment by llorca** [comment]
To fix the corner radius clipping caused by `justify-contents` on `.bp-table-cell.pt-loading`, we can try using an alternate way of centering the skeleton in the cell. Try removing `justify-contents` from `.bp-table-cell.pt-loading`, then add the following to `.bp-table-cell .pt-skeleton`:
```scss
position: absolute;
top: 50%;
transform: translateX(-50%);
```

File: packages/table/src/common/_loading.scss [hunk 4]
```diff
@@ -14,25 +14,39 @@
...
+  // stylelint-disable-next-line declaration-no-important
+  color: transparent !important;
+}
+
+@mixin disable-interaction() {
+  cursor: default;
+  pointer-events: none;
+}
+
+@mixin vertical-align() {
```
Comment (giladgray) [inline packages/table/src/common/_loading.scss]: thing is, `vertical-align` is an actual CSS property which this mixin has nothing to do with.

`vertical-center`?

File: packages/table/src/common/_loading.scss [hunk 5]
```diff
@@ -14,25 +14,39 @@
...
+  display: flex;
+  flex-direction: column;
+  justify-content: center;
+}
 
+@mixin table-skeleton($skeleton-height) {
   .pt-skeleton {
     opacity: 0;
-    height: $pt-grid-size / 2;
+    height: $skeleton-height * 1px;
```
Comment (giladgray) [inline packages/table/src/common/_loading.scss]: _whoa_ please just expect a unit from the variable. if the variable is called `height` then it should be usable as a `height` (ie, has a unit).

File: packages/table/src/common/_loading.scss [hunk 6]
```diff
@@ -14,25 +14,39 @@
...
     animation: $cell-transition-duration linear forwards skeleton-fade-in, $skeleton-animation;
-    animation-delay: 0, $cell-transition-duration;
+    animation-delay: 0s, $cell-transition-duration;
+  }
+}
+
+@mixin cell-loading {
```
Comment (giladgray) [inline packages/table/src/common/_loading.scss]: this mixin is used exactly once (in _cell.scss) but looks very similar to other instances of loading cells. suggests a potential refactor, though i'm not sure what...

File: packages/table/src/common/_loading.scss (hunk 4 above)
Comment (michael-yx-wu) [inline packages/table/src/common/_loading.scss]: you're right that's confusing. renaming!

**Comment by michael-yx-wu** [comment]
This is ready for review again 😄

File: packages/table/src/common/_loading.scss [hunk 7]
```diff
@@ -14,25 +14,29 @@
...
+@mixin disable-interaction() {
+  cursor: default;
+  pointer-events: none;
+}
+
+@mixin vertical-center() {
+  display: flex;
+  flex-direction: column;
+  justify-content: center;
+}
```
Comment (giladgray) [inline packages/table/src/common/_loading.scss]: hmm my primary hesitation with these mixins is that they have nothing to do with table loading. they are a) more generally useful but b) not complex or interesting. kinda suggests they're unnecessary overhead and we'd be better off without them in this PR.

File: packages/table/src/common/_loading.scss [hunk 8]
```diff
@@ -14,25 +14,29 @@
...
 
+@mixin table-skeleton($skeleton-height) {
```
Comment (giladgray) [inline packages/table/src/common/_loading.scss]: again, this mixin is not table specific but would be useful as a way to customize the appearance of a skeleton.

however, how useful are these styles? given that only `height` changes, do we really need to set the other three properties every time or can we set them once on `.pt-skeleton` and not have a mixin?

File: packages/table/src/headers/_headers.scss [hunk 9]
```diff
@@ -47,6 +49,12 @@
       rgba($header-selected-background-color, 0.1)
     );
   }
+
+  &.pt-loading {
+    @include disable-interaction();
+    display: flex;
+    flex-direction: column;
```
Comment (giladgray) [inline packages/table/src/headers/_headers.scss]: see here's `vertical-center()` minus one line so you can't use the mixin 😢 
may as well not have the mixin at all.

File: packages/table/src/headers/_headers.scss [hunk 10]
```diff
@@ -168,10 +175,7 @@
 
 .bp-table-column-name-text {
   padding: $cell-padding;
-}
-
-.bp-table-row-name-text {
```
Comment (giladgray) [inline packages/table/src/headers/_headers.scss]: seems this selector has been removed. is that safe?

File: packages/table/src/headers/columnHeaderCell.tsx [hunk 11]
```diff
@@ -145,21 +159,26 @@ export class ColumnHeaderCell extends React.Component<IColumnHeaderCellProps, IC
     }
 
     private renderName() {
-        const { useInteractionBar, name, renderName } = this.props;
+        const { loading, useInteractionBar, name, renderName } = this.props;
```
Comment (giladgray) [inline packages/table/src/headers/columnHeaderCell.tsx]: alphabetize real quick

File: packages/table/src/headers/rowHeaderCell.tsx [hunk 12]
```diff
@@ -28,6 +31,15 @@ export interface IRowHeaderCellProps extends IProps {
     name?: string;
 
     /**
+     * If true, the row name (string or `ReactElement`) will be replaced with a fixed-height
```
Comment (giladgray) [inline packages/table/src/headers/rowHeaderCell.tsx]: remove parenthetical. see corresponding prop docs in file above where you did a great job.

File: packages/table/src/headers/rowHeaderCell.tsx [hunk 13]
```diff
@@ -56,21 +68,25 @@ export class RowHeaderCell extends React.Component<IRowHeaderCellProps, IRowHead
...
     public render() {
-        const { className, isActive, isRowSelected, name, resizeHandle, style } = this.props;
-        const classes = classNames(className, "bp-table-header", {
+        const { className, isActive, isRowSelected, loading, name, resizeHandle, style } = this.props;
+        const rowHeaderClasses = classNames(className, "bp-table-header", {
+            [Classes.LOADING]: loading,
             "bp-table-header-active": isActive || this.state.isActive,
             "bp-table-header-selected": isRowSelected,
         });
+        const rowNameClasses = classNames("bp-table-row-name", { [Classes.LOADING]: loading });
```
Comment (giladgray) [inline packages/table/src/headers/rowHeaderCell.tsx]: why do we need `LOADING` on row header and row name?

Comment (michael-yx-wu) [inline packages/table/src/headers/rowHeaderCell.tsx]: i put loading class on the header cell so that i can conditionally modify the styles of the row name child element. the headers are a bit complicated in that unlike cells, they hold a name and potentially other arbitrary content. the additional loading class on the row name child element itself is to prevent touching the styles of the other content

Comment (giladgray) [inline packages/table/src/headers/rowHeaderCell.tsx]: but can't you do `.pt-loading .bp-table-row-name { styles }`? put the C in CSS

File: packages/table/src/common/_loading.scss (hunk 8 above)
Comment (michael-yx-wu) [inline packages/table/src/common/_loading.scss]: because headers accept arbitrary children, it's possible that users will use the `.pt-skeleton` class on children. is it okay to be applying additional styles to this class under their noses? if so, we could stick these additional properties on `.bp-table-header .pt-skeleton`

File: packages/table/src/common/_loading.scss [hunk 14]
```diff
@@ -14,25 +14,29 @@
...
   .pt-skeleton {
     opacity: 0;
-    height: $pt-grid-size / 2;
+    height: $skeleton-height;
     animation: $cell-transition-duration linear forwards skeleton-fade-in, $skeleton-animation;
-    animation-delay: 0, $cell-transition-duration;
+    animation-delay: 0s, $cell-transition-duration;
```
Comment (michael-yx-wu) [inline packages/table/src/common/_loading.scss]: this syntax doesn't seem to fully work in firefox for some reason: the fade in occurs but the glow animation never happens. perhaps an implementation bug on their end?

File: packages/table/src/cell/_cell.scss [hunk 15]
```diff
@@ -5,17 +5,19 @@
 @import "./common";
 
 .bp-table-cell {
-  @include cell-cursors();
-}
-
-.bp-table-cell,
-.bt-table-row-header-cell {
```
Comment (michael-yx-wu) [inline packages/table/src/cell/_cell.scss]: this is spelled wrong (`.bt`) and isn't used anymore

File: packages/table/src/headers/_headers.scss [hunk 16]
```diff
@@ -215,17 +217,41 @@
...
   display: flex;
   flex-direction: column;
   text-align: left;
   font-size: $pt-font-size;
+
+  &.pt-loading {
+    .bp-table-column-name-text {
+      @include table-skeleton($pt-grid-size * 0.7);
+      @include vertical-center();
+      padding: ($pt-grid-size * 1.1) $pt-grid-size ($pt-grid-size * 1.2);
```
Comment (michael-yx-wu) [inline packages/table/src/headers/_headers.scss]: the odd height of the table skeleton throws makes the entire shift up and down when transitioning between loading <=> not loading, hence odd padding here

File: packages/table/src/common/_loading.scss (hunk 7 above)
Comment (michael-yx-wu) [inline packages/table/src/common/_loading.scss]: Perhaps this could live somewhere else? I removed the other mixins but I'm still using this one 3 times in `_headers.scss`

**Comment by michael-yx-wu** [comment]
Cut down the lines of css and grouped all the loading styles together. It's personally a lot easier to read, but let me know if that's not the case.

File: .vscode/settings.json [hunk 17]
```diff
@@ -1,3 +1,4 @@
 {
-    "typescript.tsdk": "./node_modules/typescript/lib"
+    "typescript.tsdk": "./node_modules/typescript/lib",
+    "vsicons.presets.angular": false
```
Comment (michael-yx-wu) [inline .vscode/settings.json]: oops, will revert these changes

File: packages/table/src/headers/columnHeaderCell.tsx [hunk 18]
```diff
@@ -59,6 +63,14 @@ export interface IColumnHeaderCellProps extends IColumnNameProps, IProps {
     isColumnSelected?: boolean;
 
     /**
+     * If true, the column `name` will be replaced with a fixed-height skeleton and the
+     * `resizeHandle` will not be rendered. If passing in additional children to this component, you
+     * will also want to conditionally apply the `.pt-skeleton` class where appropriate.
+     * @default false
+     */
+    loading?: boolean;
```
Comment (gscshoyru) [inline packages/table/src/headers/columnHeaderCell.tsx]: Asking because new to all of this and not sure about the styles -- all of the other boolean properties here are seem to start with "is", this one doesn't. Is there a reason for this?

File: packages/table/src/headers/columnHeaderCell.tsx [hunk 19]
```diff
@@ -145,13 +159,17 @@ export class ColumnHeaderCell extends React.Component<IColumnHeaderCellProps, IC
...
-        const nameComponent = (renderName == null) ? defaultName : renderName(name);
+        const defaultName = <div className="bp-table-truncated-text">{name}</div>;
+        const nameComponent = (
+            <LoadableContent loading={loading} variableLength={true}>
+                {renderName == null ? defaultName : renderName(name)}
+            </LoadableContent>
+        );
         if (useInteractionBar) {
             return (
-                <div className={HEADER_COLUMN_NAME_CLASSNAME} title={name}>
```
Comment (gscshoyru) [inline packages/table/src/headers/columnHeaderCell.tsx]: Why? It's not removed below; why is it removed here?

**Comment by adidahiya** [comment]
> webpack errors on circle (these don't fail the build but should be fixed)

😱 😱 🎏

**Comment by michael-yx-wu** [comment]
@adidahiya The specific webpack @giladgray mentioned have been fixed. They slipped through before because this PR was started before we had consistent tsconfigs across BP packages 😨

**Comment by adidahiya** [comment]
@michael-yx-wu my concern was about them _not failing the build_

File: packages/table/src/headers/columnHeaderCell.tsx (hunk 18 above)
Comment (michael-yx-wu) [inline packages/table/src/headers/columnHeaderCell.tsx]: it actually used to be `isLoading`! @themadcreator suggested i change it to `loading`. this is consistent with the blueprint button `loading` prop

File: packages/table/src/headers/columnHeaderCell.tsx (hunk 19 above)
Comment (michael-yx-wu) [inline packages/table/src/headers/columnHeaderCell.tsx]: oops good catch

File: packages/table/src/common/_loading.scss [hunk 20]
```diff
@@ -14,25 +14,17 @@
...
+  display: flex;
+  flex-direction: column;
+  justify-content: center;
+}
 
+@mixin table-skeleton() {
   .pt-skeleton {
     opacity: 0;
-    height: $pt-grid-size / 2;
+    height: $pt-grid-size * 0.5;
```
Comment (giladgray) [inline packages/table/src/common/_loading.scss]: i much prefer `/ 2` to `* 0.5`. why this change?

File: packages/table/src/headers/_headers.scss [hunk 21]
```diff
@@ -254,3 +252,30 @@
...
+  @include vertical-center();
+
+  .bp-table-column-name,
+  .bp-table-row-name {
+    flex: 1;
+  }
+
+  .bp-table-column-name-text {
+    @include vertical-center();
+    padding: ($pt-grid-size * 1.1) $pt-grid-size ($pt-grid-size * 1.2);
```
Comment (giladgray) [inline packages/table/src/headers/_headers.scss]: _whoa_ why the freaky padding? 

i'd expect `vertical-center` to center correctly, to not require 1px adjustments.

File: packages/table/src/headers/_headers.scss [hunk 22]
```diff
@@ -254,3 +252,30 @@
...
+
+    .pt-skeleton {
+      height: $pt-grid-size * 0.7;
```
Comment (giladgray) [inline packages/table/src/headers/_headers.scss]: this would make a good variable. `$header-skeleton-height`

File: packages/table/src/headers/columnHeaderCell.tsx (hunk 18 above)
Comment (giladgray) [inline packages/table/src/headers/columnHeaderCell.tsx]: huh awkward that it's inconsistent with the rest of the component 😢

File: packages/table/src/cell/_common.scss [hunk 23]
```diff
@@ -29,7 +29,7 @@ $cell-transition-duration: $pt-transition-duration * 3;
   transition: color $cell-transition-duration;
 
   .pt-dark & {
-    &:not([class*="pt-intent-"]) {
+    &:not([class*="pt-intent-"]):not(.pt-loading) {
```
Comment (llorca) [inline packages/table/src/cell/_common.scss]: Isn't `&:not([class*="pt-intent-"], .pt-loading)` valid?

File: packages/table/src/headers/columnHeaderCell.tsx (hunk 18 above)
Comment (michael-yx-wu) [inline packages/table/src/headers/columnHeaderCell.tsx]: i'm indifferent, but should i change to `isLoading`?

File: packages/table/src/headers/_headers.scss (hunk 21 above)
Comment (michael-yx-wu) [inline packages/table/src/headers/_headers.scss]: it does center the children content of `.bp-table-column-name-text`. the default header is 30px high. the problem with a 7px height skeleton is that we need unequal top / bottom padding to make it 30px (i.e. 11px top, 22px bottom)

File: packages/table/src/common/_loading.scss (hunk 7 above)
Comment (llorca) [inline packages/table/src/common/_loading.scss]: It's a generically named mixin that's only used in one file. I tend to agree that the complexity isn't worth such an abstract mixin

File: packages/table/src/headers/columnHeaderCell.tsx (hunk 18 above)
Comment (llorca) [inline packages/table/src/headers/columnHeaderCell.tsx]: We picked `loading` earlier, we should stick with it everywhere. Otherwise, file an issue to change all instances of `loading` to `isLoading`. What matters here is consistency in naming

File: packages/table/src/headers/_headers.scss (hunk 21 above)
Comment (giladgray) [inline packages/table/src/headers/_headers.scss]: ok gotcha. in that case, would be even better to compute these values from the variable i proposed below. use `floor()` and `ceil()` functions to ensure whole pixel values.

Comment (giladgray) [inline packages/table/src/headers/_headers.scss]: also, what's wrong with 8px instead of 7?

File: packages/table/src/common/_loading.scss (hunk 7 above)
Comment (michael-yx-wu) [inline packages/table/src/common/_loading.scss]: should i move this to `_headers.scss`? I don't want to repeat the content in this mixin

File: packages/table/src/headers/columnHeaderCell.tsx (hunk 18 above)
Comment (michael-yx-wu) [inline packages/table/src/headers/columnHeaderCell.tsx]: i'll leave it as is. perhaps in 2.0 we change `is[PropertyName]` type variables to just `[PropertyName]`.

File: packages/table/src/common/_loading.scss (hunk 7 above)
Comment (llorca) [inline packages/table/src/common/_loading.scss]: If it's not used outside of `_loading.scss`, why should it move to `_headers.scss`?

File: packages/table/src/headers/_headers.scss (hunk 21 above)
Comment (michael-yx-wu) [inline packages/table/src/headers/_headers.scss]: the designers were against 8px, too thick. there was a bushy eyebrow meme going on for a while, because that's what it looks like 😛.

File: packages/table/src/cell/_common.scss (hunk 23 above)
Comment (michael-yx-wu) [inline packages/table/src/cell/_common.scss]: that syntax is proposed in selectors level 4: https://drafts.csswg.org/selectors-4/#negation

in selectors level 3, `:not` only takes a simple argument. so while this compiles and seems to work in chrome and FF, i'm on the latest stable version for both. i'm going to go with the more verbose syntax for now since it's guaranteed it'll work on older browsers

File: packages/table/src/common/_loading.scss (hunk 7 above)
Comment (michael-yx-wu) [inline packages/table/src/common/_loading.scss]: it's used a few times `_headers.scss`

Comment (llorca) [inline packages/table/src/common/_loading.scss]: Ok, then maybe it could live in `headers/_common.scss`, renamed to `header-vertical-center()`?

Comment (michael-yx-wu) [inline packages/table/src/common/_loading.scss]: hm, but it's also a loading specific style. is it okay for me to leave it here? i just checked and it's being used in `_cell.scss` (once) and `_header.scss` (three times)

Comment (llorca) [inline packages/table/src/common/_loading.scss]: Ok, it can stay here then if it's specific to loading. The name still doesn't work -- not clear that's related to loading. What about `skeleton-vertical-center`?

Comment (michael-yx-wu) [inline packages/table/src/common/_loading.scss]: i'll change it to `vertical-center-content` since it's always applied to the parent of skeletons

File: packages/table/src/cell/_common.scss [hunk 24]
```diff
@@ -62,3 +62,9 @@ $cell-transition-duration: $pt-transition-duration * 3;
     cursor: $select-cell-cursor;
   }
 }
+
+@mixin cell-content-align-vertical() {
```
Comment (llorca) [inline packages/table/src/cell/_common.scss]: That clears up naming confusion, and describes what it does: vertically center anything inside a cell. Not specific to loading or header cells. How's that @giladgray?

File: packages/table/src/headers/_headers.scss [hunk 25]
```diff
@@ -254,3 +252,32 @@
...
+  .bp-table-column-name,
+  .bp-table-row-name {
+    flex: 1;
+  }
+
+  .bp-table-column-name-text {
+    @include cell-content-align-vertical();
+    padding: floor(($column-header-min-height - $column-name-text-skeleton-height) / 2)
+      $pt-grid-size
+      ceil(($column-header-min-height - $column-name-text-skeleton-height) / 2);
```
Comment (giladgray) [inline packages/table/src/headers/_headers.scss]: this entire property is actually worthless because of the mixin. try removing it and notice how the skeleton becomes perfectly centered (and how it was clearly a pixel off with this padding).

i'd still suggest 8px height cuz even numbers are always better for centering.

Comment (giladgray) [inline packages/table/src/headers/_headers.scss]: alternately, just set the super simple `padding: $pt-grid-size;` to ensure space on all sides, and let flex box handle the actual centering.

Comment (michael-yx-wu) [inline packages/table/src/headers/_headers.scss]: done

File: .vscode/settings.json [hunk 26]
```diff
@@ -1,3 +1,4 @@
 {
-    "typescript.tsdk": "./node_modules/typescript/lib"
-}
\ No newline at end of file
+    "typescript.tsdk": "./node_modules/typescript/lib",
+    "vsicons.presets.angular": false
+}
```
Comment (giladgray) [inline .vscode/settings.json]: what happened to reverting this?

**Comment by michael-yx-wu** [comment]
reverted the .vscode/settings change!
